pkg/org/sso_handler.go

```diff
@@ -0,0 +1,195 @@
+package org
+
+import (
+	"context"
+	"crypto/rand"
+	"encoding/base64"
+	"errors"
+	"fmt"
+	"io"
+	"net/http"
+	"time"
+
+	"github.com/coreos/go-oidc/v3/oidc"
+	"github.com/uptrace/bunrouter"
+	"github.com/uptrace/uptrace/pkg/bunapp"
+	"github.com/uptrace/uptrace/pkg/bunconf"
+	"github.com/uptrace/uptrace/pkg/httputil"
+	"go.uber.org/zap"
+	"golang.org/x/oauth2"
+)
+
+type SSOHandler struct {
+	app  *bunapp.App
+	oidc *OIDCProvider
+}
+
+type Methods struct {
+	OIDC *OIDCMethod `json:"oidc"`
+}
+
+func NewSSOHandler(app *bunapp.App) *SSOHandler {
+	conf := app.Config()
+	ctx := app.Context()
+
+	var oidc *OIDCProvider
+	if conf.UserProviders.OIDC != nil {
+		if provider, err := NewOIDCProvider(ctx, conf.UserProviders.OIDC); provider != nil {
+			oidc = provider
+		} else if err != nil {
+			app.Zap(ctx).Error("Failed to initialize OIDC user provider", zap.Error(err))
+		}
+	}
+
+	return &SSOHandler{
+		app:  app,
+		oidc: oidc,
+	}
+}
+
+func (h *SSOHandler) ListMethods(w http.ResponseWriter, req bunrouter.Request) error {
+	app := h.app
+	ctx := req.Context()
+
+	methods := &Methods{}
+
+	if h.oidc != nil {
+		oidc, err := h.oidc.Start(w, req)
+		if err != nil {
+			app.Zap(ctx).Error("Failed to start OIDC flow", zap.Error(err))
+		} else if oidc != nil {
+			methods.OIDC = oidc
+		}
+	}
+
+	return httputil.JSON(w, methods)
+}
+
+//------------------------------------------------------------------------------
+
+type OIDCMethod struct {
+	DisplayName string `json:"name"`
+	RedirectURL string `json:"url"`
+}
+
+type OIDCProvider struct {
```

Comment:
Let's remove `OIDCProvider` altogether and just only have `SSOHandler` instead?

Reply:
I actually think the seperation is better. It isolates the OIDC-specific implementation. Supporting other flows like SAML could be useful in the future.